src/plcc/load_spec/validate_spec/validate_syntactic_spec/Grammar_test.py

```diff
@@ -0,0 +1,88 @@
+from pytest import raises, fixture
+from .Grammar import Grammar
+from .errors import InvalidFormError, InvalidNonterminalError
+
+@fixture
+def grammar():
+    return Grammar()
+
+@fixture
+def nonterminal():
+    return 'nonTerminal'
+
+@fixture
+def terminal():
+    return 'TERMINAL'
+
+def test_empty_grammar(grammar):
+    assert len(grammar.getRules()) == 0
+    assert grammar.getStartSymbol() == None
+
+def test_add_rule(grammar, nonterminal, terminal):
+    grammar.addRule(nonterminal, [terminal])
+    assert grammar.getRules() == {nonterminal: [[terminal]]}
+
+def test_is_terminal(grammar, terminal):
+    assert grammar.isTerminal(terminal) == True
+    assert grammar.isNonterminal(terminal) == False
+
+def test_is_nonterminal(grammar, nonterminal):
+    grammar.addRule(nonterminal, [])
+    assert grammar.isTerminal(nonterminal) == False
+    assert grammar.isNonterminal(nonterminal) == True
+
+def test_get_start_symbol_none(grammar):
+    assert grammar.getStartSymbol() == None
+
+def test_get_start_symbol(grammar, nonterminal):
+    grammar.addRule(nonterminal, [])
+    assert grammar.getStartSymbol() == nonterminal
+
+def test_multiple_rules_one_nonterminal(grammar, nonterminal, terminal):
+    grammar.addRule(nonterminal, [terminal])
+    grammar.addRule(nonterminal, [])
+    assert grammar.getRules() == {nonterminal: [[terminal], []]}
+
+def test_add_multiple_nonterminals(grammar, nonterminal, terminal):
+    grammar.addRule(nonterminal, [terminal])
+    anotherNonterminal = getanotherNonterminal()
+    grammar.addRule(anotherNonterminal, [])
+    assert grammar.getRules() == {nonterminal: [[terminal]], anotherNonterminal: [[]]}
+
+def test_add_terminal_and_nonterminal_to_sets(grammar, nonterminal, terminal):
+    anotherNonterminal = getanotherNonterminal()
+    grammar.addRule(anotherNonterminal, [])
+    grammar.addRule(nonterminal, [terminal, anotherNonterminal])
+    assert grammar.getRules() == {anotherNonterminal: [[]], nonterminal: [[terminal, anotherNonterminal]]}
+    assert len(grammar.getNonterminals()) == 2
+    assert len(grammar.getTerminals()) == 1
+    assert nonterminal in grammar.getNonterminals()
+    assert anotherNonterminal in grammar.getNonterminals()
+    assert terminal in grammar.getTerminals()
+
+def test_no_duplicate_terminals(grammar, nonterminal, terminal):
+    grammar.addRule(nonterminal, [terminal])
+    grammar.addRule(nonterminal, [terminal])
+    assert len(grammar.getTerminals()) == 1
+    assert terminal in grammar.getTerminals()
+
+def test_no_duplicate_nonterminals(grammar, nonterminal, terminal):
+    grammar.addRule(nonterminal, [terminal])
+    grammar.addRule(nonterminal, [terminal])
+    assert len(grammar.getNonterminals()) == 1
+    assert nonterminal in grammar.getNonterminals()
+
+def test_invalid_nonterminal_throws_invalid_nonterminal_error(grammar, terminal):
```

Comment:
In Grammar's universe, a valid symbol (terminal or nonterminal) is any object that words with Python's set data structure.

Otherwise it's impossible to have an invalid terminal or nonterminal. Any symbol that appears on the LHS of a rule is a nonterminal, all other symbols are terminals.

While you are at it, try to generalize the type of symbols to any object that works with sets. If there isn't a single type that does this, then generalize to object (I think that is Any in the type hints) and then check that they have the necessary methods (__eq__ and __hash__ ?). And that should be all Grammar cares about.